docs/criteria.md

```diff
@@ -0,0 +1,877 @@
+
+It has two essential purposes:
+
+1. It is used to distinguish criteria
+2. When performing matching, the rule is converted into a [camel case](https://en.wikipedia.org/wiki/Camel_case) name of a method that handles this criterion (see ["How criteria works"](#how-criteria-works)).
+
+### Data
+
+It is just an optional array of additional criterion data. For example, "User has posted at least X messages" criterion has a data array with one element: a number of messages.
+
+## How criteria system works
+
+In this sections, we describe how criteria system works from A to Z.
+
+### Template
+
+It all starts from template code. Here is how criteria look inside templates:
+
+```html
+<xf:checkbox label="Criteria container">
+    
+    <!-- Criterion -->
+    <xf:option name="foo_criteria[criterion_1_rule][rule]" value="criterion_1_rule" ... />
+    
+    <!-- Criterion with data -->
+    <xf:option name="bar_criteria[criterion_2_rule][rule]" value="criterion_2_rule" ... >
+        <xf:... name="bar_criteria[criterion_2_rule][data][param_1]" ... />
+        <xf:... name="bar_criteria[criterion_2_rule][data][param_2]" ... />
+    </xf:option>
+        
+</xf:checkbox>
+```
+
+As you can see, criterion is simply a checkbox with optional input fields inside (criterion data). Let's analyze the code:
+
+-   `foo_criteria` and `bar_criteria` are the input containers and usually `foo` and `bar` parts refer to criteria type. For example, `user_criteria[...]` lets us know that this criteria belong to User criteria.
+-   `value="criterion_1_rule"` and `value="criterion_2_rule"` are, obviously, the rules of criteria.
+
+!!! note
+    Keep in mind that `criterion_1/2_rule` in `name` attributes may not have to be criteria rules! These are just names for input containers. You can easily write `<xf:option name="foo[bar][rule]" value="criterion_rule" />` and it will work correctly. The criterion rule will be `criterion_rule`, not `bar`.
+
+### (Optionally) Storing selected criteria
+
+Inside the controller, the criteria form data from the previous section can be filtered, serialized and saved in database columns of `mediumblob` type for better days:
+
+```php
+$fooCriteriaInput = $this->filter('foo_criteria', 'array');
+$barCriteriaInput = $this->filter('bar_criteria', 'array');
+
+$form->basicEntitySave($bazEntity, [$fooCriteriaInput, $barCriteriaInput]);
+```
+
+The example `$bazEntity` structure:
+
+```php
+public static function getStructure(Structure $structure)
+{
+    $structure->table = 'xf_baz';
+    $structure->shortName = 'XF:Baz';
+    $structure->primaryKey = 'baz_id';
+    $structure->columns = [
+        'baz_id' => ['type' => self::UINT, 'autoIncrement' => true],
+        'foo_criteria' => ['type' => self::SERIALIZED_ARRAY, 'default' => [], 'required' => 'please_select_criteria_that_must_be_met'],
+        'bar_criteria' => ['type' => self::SERIALIZED_ARRAY, 'default' => []]
```

Comment:
This is sort of future proofing, though it would work in 2.0 as well.

In 2.1, most serialized arrays have been turned into JSON, including where we store criteria. SERIALIZED_ARRAY is generally deprecated (SERIALIZED is fine if you really need it), so I think it's probably best to set these to JSON_ARRAY to match best practices.

Reply:
Is `JSON_ARRAY` also similar deprecated?